docs/source/dataframe-design.rst

```diff
@@ -0,0 +1,111 @@
+Internal Design
+===============
+
+Dask dataframes coordinate many Pandas arrays arranged along an index. We
+define a ``dask.dataframe`` object with the following components:
+
+- A dask graph with a special set of keys designating partitions, such as
+  ``('x', 0), ('x', 1), ...``.
+- A name to identify which keys in the dask graph refer to this dataframe, such
+  as ``'x'``.
+- An empty pandas object containing appropriate metadata (e.g.  column names,
+  dtypes, etc...).
+- A sequence of partition boundaries along the index, called ``divisions``.
+
+Metadata
+--------
+
+Many dataframe operations rely on knowing the name and dtype of columns. To
+keep track of this information, all ``dask.dataframe`` objects have a ``_meta``
+attribute which contains an empty pandas object with the same dtypes and names.
+For example:
+
+.. code-block:: python
+
+   >>> df = pd.DataFrame({'a': [1, 2, 3], 'b': ['x', 'y', 'z']})
+   >>> ddf = dd.from_pandas(df, npartitions=2)
+   >>> ddf._meta
+   Empty DataFrame
+   Columns: [a, b]
+   Index: []
+   >>> ddf._meta.dtypes
+   a     int64
+   b    object
+   dtype: object
+
+Internally ``dask.dataframe`` does its best to propagate this information
+through all operations, so most of the time a user shouldn't have to worry
+about this.  Usually this is done by evaluating the operation on a small sample
+of fake data, which can be found on the ``_meta_nonempty`` attribute:
+
+.. code-block:: python
+
+   >>> ddf._meta_nonempty
+      a    b
+   0  1  foo
+   1  1  foo
+
+Sometimes this operation may fail in user defined functions (e.g. when using
+``DataFrame.apply``), or may be prohibitively expensive.  For these cases, many
+functions support an optionally ``meta`` keyword, which allows specifying the
+metadata directly, avoiding the the inference step. For convenience, this
+supports several options:
+
+- A pandas object with appropriate dtypes and names. If not empty, an empty
+  slice will be taken:
+
+.. code-block:: python
+
+  >>> ddf.map_partitions(foo, meta=pd.DataFrame({'a': [1], 'b': [2]}))
+
+- A description of the appropriate names and dtypes. This can take several forms:
+
+    * A ``dict`` of ``{name: dtype}`` or an iterable of ``(name, dtype)``
```

Comment:
This renders as code, at least in github's renderer.  Might have to build docs locally to check.